### PR TITLE
CTPPS reco update to use LHCInfoPer* records

### PR DESCRIPTION
**Pull request by Glitchmin**
Following LHCInfo update introducing 2 new records `LHCInfoPerFill` and `LHCInfoPerLS` (https://github.com/cms-sw/cmssw/pull/39495) and their O2Os (https://github.com/cms-sw/cmssw/pull/40817) this PR introduces:
* helper `LHCInfoCombined` class that is meant to be used where old `LHCInfo` was used 
* every change needed to run `recoCTPPS` with only `LHCInfoPer*`

LHCInfoCombined can be created using either `LHCInfoPer*` or `LHCInfo` and is not meant to be serialised. It selects the source depending on the constructor parameter `useNewLHCInfo` that can be set based on era (since Run3 GTs won't contain `LHCInfo` records and Run2 GTs contain only those) in the python _cff file, eg.:

https://github.com/CTPPS/cmssw/blob/bb215fd85be7746ec3449516b4c0565aaf1f624f/CalibPPS/ESProducers/python/ctppsOpticalFunctions_cff.py#L32-L33

In order to make `recoCTPPS` able to use `LHCInfoPer*` or `LHCInfo` following modules were updated:
* CTPPSProtonProducer
* CTPPSInterpolatedOpticalFunctionsESSource

From PPS side this is the version of code meant to be used in re-reco of 2022 data.

#### PR validation:

Validation can be done by running `recoCTPPS` for fill 9019:
```
cd Calibration/PPSAlCaRecoProducer/test
cmsRun 2023_lhcinfo_test_recoCTPPS_cfg.py
```
and checking if h_multiplicity, h_xi and p_th_y_vs_xi for singleRP plots in `alcareco_protons_express.root` are not empty.

Additionally it was tested by comparing the aformentioned plots with the same plots generated with the code before changes from this PR. To generate plots for code before changes 2023_lhcinfo_test_recoCTPPS_cfg.py was copied and run with a tag with properly generated LHCInfo payloads for fill 9019 ( 130X_dataRun3_Prompt_forLHCInfo_Candidate_2023_07_28_12_52_37 )

### CODE REVIEW

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by Glitchmin** [comment]
I added a commit addressing build warnings, this commit will be squashed once the reviewing is done

**Comment by mandrenguyen** [comment]
type ctpps

**Comment by francescobrivio** [comment]
@cmsbuild please test

 - even though final tests can come only with an updated GT we can start see how the code is looking

**Comment by francescobrivio** [comment]
Ok all the failures observed are expected (relvals and AddOn tests) and related to the fact that the new records are missing from the GTs. Me and @JanChyczynski are working on this:
 - For data, the tags are already available (see this [CMSTalk post](https://cms-talk.web.cern.ch/t/online-prompt-new-lhcinfoper-tags-for-run3/28208))
 - For MC, the payloads are being prepared


Only exception is the failure in the unitTests which should not be related to this Pm, but is being investigated nonetheless.

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc [hunk 1]
```diff
@@ -28,19 +29,26 @@ class CTPPSInterpolatedOpticalFunctionsESSource : public edm::ESProducer {
...
 CTPPSInterpolatedOpticalFunctionsESSource::CTPPSInterpolatedOpticalFunctionsESSource(const edm::ParameterSet &iConfig)
-    : currentCrossingAngle_(-1.), currentDataValid_(false) {
+    : currentCrossingAngle_(-1.),
+      currentDataValid_(false),
+      useNewLHCInfo_(iConfig.getParameter<bool>("useNewLHCInfo")) {
   auto cc = setWhatProduced(this, iConfig.getParameter<std::string>("opticsLabel"));
   opticsToken_ = cc.consumes(edm::ESInputTag("", iConfig.getParameter<std::string>("opticsLabel")));
   lhcInfoToken_ = cc.consumes(edm::ESInputTag("", iConfig.getParameter<std::string>("lhcInfoLabel")));
+  lhcInfoPerLSToken_ = cc.consumes(edm::ESInputTag("", iConfig.getParameter<std::string>("lhcInfoPerLSLabel")));
+  lhcInfoPerFillToken_ = cc.consumes(edm::ESInputTag("", iConfig.getParameter<std::string>("lhcInfoPerFillLabel")));
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:51]: I think these tokens could be moved to the init list, and have them `const`-s, is that not the case?

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc [hunk 2]
```diff
@@ -28,19 +29,26 @@ class CTPPSInterpolatedOpticalFunctionsESSource : public edm::ESProducer {
 private:
   edm::ESGetToken<LHCOpticalFunctionsSetCollection, CTPPSOpticsRcd> opticsToken_;
   edm::ESGetToken<LHCInfo, LHCInfoRcd> lhcInfoToken_;
+  edm::ESGetToken<LHCInfoPerLS, LHCInfoPerLSRcd> lhcInfoPerLSToken_;
+  edm::ESGetToken<LHCInfoPerFill, LHCInfoPerFillRcd> lhcInfoPerFillToken_;
```
Comment (tvami) [inline CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:33]: if my comment bellow works, these are to be `const`-s

File: CalibPPS/ESProducers/python/ctppsLHCInfo_cff.py [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 # by default, LHCInfo is now loaded from CondDB using a GT
 ctppsLHCInfoLabel = cms.string("")
+ctppsLHCInfoPerLSLabel = cms.string("")
+ctppsLHCInfoPerFillLabel = cms.string("")
```
Comment (tvami) [inline CalibPPS/ESProducers/python/ctppsLHCInfo_cff.py:6]: why are these needed? the `fillDesc` makes sure there is a default of `""` so none of these should be needed

File: Calibration/PPSAlCaRecoProducer/test/2022_lhcinfo_test_AlCaRecoProducer.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+function die { echo $1: status $2; exit $2; }
+
+customise_commands="process.GlobalTag.toGet = cms.VPSet()\
```
Comment (tvami) [inline Calibration/PPSAlCaRecoProducer/test/2022_lhcinfo_test_AlCaRecoProducer.sh:4]: these should be taken from the DB (after Wagner uploads them), and I guess not just the DB but through a GT

File: RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc [hunk 5]
```diff
@@ -73,6 +81,8 @@ class CTPPSProtonProducer : public edm::stream::EDProducer<> {
   bool opticsValid_;
   edm::ESWatcher<CTPPSInterpolatedOpticsRcd> opticsWatcher_;
 
+  edm::ESGetToken<LHCInfoPerLS, LHCInfoPerLSRcd> lhcInfoPerLSToken_;
+  edm::ESGetToken<LHCInfoPerFill, LHCInfoPerFillRcd> lhcInfoPerFillToken_;
   edm::ESGetToken<LHCInfo, LHCInfoRcd> lhcInfoToken_;
```
Comment (tvami) [inline RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc:86]: in this case these token are in the init list, so could be `const`

**Comment by francescobrivio** [comment]
@JanChyczynski @Glitchmin did you see Tamas's last review comments?

**Comment by francescobrivio** [comment]
test parameters:
 - pull_request = #42551

**Comment by francescobrivio** [comment]
@cmsbuild please test

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc (hunk 1 above)
Comment (mmusich) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:51]: I don't think that's possible, when the `ConsumesCollector` is an object instantiated from within the class constructor ([dxr](https://cmssdt.cern.ch/dxr/CMSSW/search?q=cc.consumes&case=true&offset=127)).

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc (hunk 2 above)
Comment (mmusich) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:33]: See https://github.com/cms-sw/cmssw/pull/42515#discussion_r1291759585

File: Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py [hunk 6]
```diff
@@ -14,9 +14,9 @@
 from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import totemTimingRawToDigi as _totemTimingRawToDigi
 from EventFilter.CTPPSRawToDigi.ctppsRawToDigi_cff import ctppsPixelDigis as _ctppsPixelDigis
 
-ctppsDiamondRawToDigiAlCaRecoProducer  = _ctppsDiamondRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
-totemTimingRawToDigiAlCaRecoProducer = _totemTimingRawToDigi.clone(rawDataTag = 'hltPPSCalibrationRaw')
-ctppsPixelDigisAlCaRecoProducer = _ctppsPixelDigis.clone(inputLabel = 'hltPPSCalibrationRaw')
+ctppsDiamondRawToDigiAlCaRecoProducer  = _ctppsDiamondRawToDigi.clone(rawDataTag = 'rawDataCollector')
+totemTimingRawToDigiAlCaRecoProducer = _totemTimingRawToDigi.clone(rawDataTag = 'rawDataCollector')
+ctppsPixelDigisAlCaRecoProducer = _ctppsPixelDigis.clone(inputLabel = 'rawDataCollector')
```
Comment (francescobrivio) [inline Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py:19]: Why where these changed from `hltPPSCalibrationRaw` to `rawDataCollector`?
This is now causing the failures observed in the Relvals-input for `1043.0,1044.0,1045.0`
And it's indeed spotted also by the (failing) PPS unitTests:
   - Calibration/PPSAlCaRecoProducer/test/testExpressPPSAlCaRecoProducer
   - Calibration/PPSAlCaRecoProducer/test/testPromptPPSAlCaRecoProducer

File: CalibPPS/ESProducers/python/ctppsLHCInfo_cff.py (hunk 3 above)
Comment (Glitchmin) [reply CalibPPS/ESProducers/python/ctppsLHCInfo_cff.py]: addressed in commit https://github.com/cms-sw/cmssw/pull/42515/commits/fd20a63bde3d3fcaa1deb60e6fc1b267ad413ec7, it will be squashed after testing and addressing comments

File: RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc (hunk 5 above)
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc]: Addressed in https://github.com/cms-sw/cmssw/pull/42515/commits/cd1fac4534f2fd682e4affde360e8f0344ec3cad

File: Calibration/PPSAlCaRecoProducer/test/2022_lhcinfo_test_AlCaRecoProducer.sh (hunk 4 above)
Comment (Glitchmin) [reply Calibration/PPSAlCaRecoProducer/test/2022_lhcinfo_test_AlCaRecoProducer.sh]: this test was temporary and was removed in https://github.com/cms-sw/cmssw/pull/42515/commits/0d1170c5b303f20b1b06568b994a27f55a1d017b since 2023_lhcinfo_test_recoCTPPS_cfg.py performs a similar task. The afformentioned commit will be squashed once the pr is tested

File: Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py (hunk 6 above)
Comment (Glitchmin) [reply Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py]: it was a change needed for a temporary test, it is fixed in https://github.com/cms-sw/cmssw/pull/42515/commits/0d1170c5b303f20b1b06568b994a27f55a1d017b

**Comment by francescobrivio** [comment]
@cmsbuild please test

File: CalibPPS/ESProducers/python/ctppsLHCInfo_cff.py [hunk 7]
```diff
@@ -1,7 +1,6 @@
 import FWCore.ParameterSet.Config as cms
```
Comment (tvami) [inline CalibPPS/ESProducers/python/ctppsLHCInfo_cff.py:1]: my suggestion would be to remove this file completely and whatever used it could take the fillDescription based auto-generated cfi file

Comment (Glitchmin) [reply CalibPPS/ESProducers/python/ctppsLHCInfo_cff.py]: addressed in 898bf11e0f06508b452854a518f42f9fbb151882

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by Glitchmin** [comment]
I know why TestCTPPSDirectProtonSimulation is failing - `Validation/CTPPS/plugins/CTPPSLHCInfoPlotter.cc` is using old `LHCInfo` and those records are not present in GTs of run3
This plotter is part of the local simulation, not related to reconstruction, I will start to work on that

File: Validation/CTPPS/plugins/CTPPSLHCInfoPlotter.cc [hunk 8]
```diff
@@ -30,7 +29,10 @@ class CTPPSLHCInfoPlotter : public edm::one::EDAnalyzer<> {
   void analyze(const edm::Event &, const edm::EventSetup &) override;
   void endJob() override;
 
-  edm::ESGetToken<LHCInfo, LHCInfoRcd> lhcInfoESToken_;
+  edm::ESGetToken<LHCInfo, LHCInfoRcd> lhcInfoToken_;
+  edm::ESGetToken<LHCInfoPerLS, LHCInfoPerLSRcd> lhcInfoPerLSToken_;
+  edm::ESGetToken<LHCInfoPerFill, LHCInfoPerFillRcd> lhcInfoPerFillToken_;
```
Comment (tvami) [inline Validation/CTPPS/plugins/CTPPSLHCInfoPlotter.cc:34]: These could be `const`-s too

**Comment by Glitchmin** [comment]
6bacb852418259a427a2ef984c5f49619f3b7889 commit is a fix to allow TestCTPPSDirectProtonSimulation run without crushing - its output was not validated

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc [hunk 9]
```diff
@@ -61,14 +72,17 @@ std::shared_ptr<LHCInterpolatedOpticalFunctionsSetCollection> CTPPSInterpolatedO
...
+      iRecord, lhcInfoPerLSToken_, lhcInfoPerFillToken_, lhcInfoToken_, useNewLHCInfo_);
 
   // is there anything to do?
-  if (currentDataValid_ && lhcInfo.crossingAngle() == currentCrossingAngle_)
+  if (currentDataValid_ && lhcInfoCombined.crossingAngle() == currentCrossingAngle_)
     return currentData_;
 
   // is crossing angle reasonable (LHCInfo is correctly filled in DB)?
-  if (lhcInfo.crossingAngle() == 0.) {
+  if (lhcInfoCombined.crossingAngle() == -1.) {
```
[the remark this answers is not in the document]
Comment (vavati) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:85]: "-1' is written in PPS online DB when there is a DIP communication failure.....

[the remark this answers is not in the document]
Comment (forthommel) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:85]: Right! See my suggestion at https://github.com/cms-sw/cmssw/pull/42515#discussion_r1295094755

File: Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py [hunk 10]
```diff
@@ -85,7 +85,8 @@
 )
 
 from RecoPPS.ProtonReconstruction.ctppsProtons_cff import ctppsProtons as _ctppsProtons
-ctppsProtonsAlCaRecoProducer = _ctppsProtons.clone(tagLocalTrackLite = 'ctppsLocalTrackLiteProducerAlCaRecoProducer')
+ctppsProtonsAlCaRecoProducer = _ctppsProtons.clone(tagLocalTrackLite = 'ctppsLocalTrackLiteProducerAlCaRecoProducer',
+verbosity = cms.untracked.uint32(2))
```
[the remark this answers is not in the document]
Comment (grzanka) [reply Calibration/PPSAlCaRecoProducer/python/ALCARECOPPSCalMaxTracks_cff.py]: I don't see that in recent commits, all the AlCaRecoProducer changes are now gone.

File: RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc [hunk 11]
```diff
@@ -51,9 +56,12 @@ class CTPPSProtonProducer : public edm::stream::EDProducer<> {
 
   bool pixelDiscardBXShiftedTracks_;
 
+  std::string lhcInfoPerLSLabel_;
```
[the remark this answers is not in the document]
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc]: Addressed in 0e2d5779cc7463948bcabf9b25bf592663fa031b

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (Glitchmin) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc]: Addressed in 62da8bb8008b690aea0da493a59d93662c4623a7

File: CondTools/RunInfo/src/LHCInfoCombined.cc [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+  crossingAngleX = infoPerLS.crossingAngleX();
+  crossingAngleY = infoPerLS.crossingAngleY();
+  betaStarX = infoPerLS.betaStarX();
+  betaStarY = infoPerLS.betaStarY();
+}
+void LHCInfoCombined::setFromPerFill(const LHCInfoPerFill& infoPerFill) { energy = infoPerFill.energy(); }
+
+float LHCInfoCombined::crossingAngle() {
+  if (crossingAngleX == 0. && crossingAngleY == 0.) {
+    return -1.;
```
[the remark this answers is not in the document]
Comment (Glitchmin) [reply CondTools/RunInfo/src/LHCInfoCombined.cc]: Addressed in 62da8bb8008b690aea0da493a59d93662c4623a7

**Comment by Glitchmin** [comment]
> > I know why TestCTPPSDirectProtonSimulation is failing - `Validation/CTPPS/plugins/CTPPSLHCInfoPlotter.cc` is using old `LHCInfo` and those records are not present in GTs of run3 This plotter is part of the local simulation, not related to reconstruction, I will start to work on that
> 
> @Glitchmin nom the Run3 GTs still contain the old LHCInfo tag, see the GT diffs in: #42551 So I'm not sure why it's failing. The [unitTest output](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d35760/34300/unitTests/src/Validation/CTPPS/test/TestCTPPSDirectProtonSimulation/testing.log) is not very descriptive.

Yes, it does make sense. Turns out that this test was crashing not because it couldn't find `LHCInfo`, in my local enviroment:
* if the newest commit is 88f740acd2b41d4681e76cf127ff83048ecafebb the test is crashing cause it can't find `LHCInfoPerLSRcd`
* after cherry-picking 6bacb852418259a427a2ef984c5f49619f3b7889 everything is fine

Maybe its worth testing this PR again? @francescobrivio @tvami 

Btw do you know which GT is this test using? Where does it get old `LHCInfo` from? I didn't find any place where GT is defined nor any place where `LHCInfo` is given directly

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc [hunk 13]
```diff
@@ -61,14 +72,17 @@ std::shared_ptr<LHCInterpolatedOpticalFunctionsSetCollection> CTPPSInterpolatedO
...
+      iRecord, lhcInfoPerLSToken_, lhcInfoPerFillToken_, lhcInfoToken_, useNewLHCInfo_);
 
   // is there anything to do?
-  if (currentDataValid_ && lhcInfo.crossingAngle() == currentCrossingAngle_)
+  if (currentDataValid_ && lhcInfoCombined.crossingAngle() == currentCrossingAngle_)
     return currentData_;
 
   // is crossing angle reasonable (LHCInfo is correctly filled in DB)?
-  if (lhcInfo.crossingAngle() == 0.) {
+  if (lhcInfoCombined.crossingAngle() == LHCInfoCombined::crossingAngleInvalid) {
```
[the remark this answers is not in the document]
Comment (Glitchmin) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:85]: I totally agree it should be a float - my mistake

File: CondTools/RunInfo/interface/LHCInfoCombined.h [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+      const edm::ESGetToken<LHCInfoPerFill, LHCInfoPerFillRcd>& tokenInfoPerFill,
+      const edm::ESGetToken<LHCInfo, LHCInfoRcd>& tokenInfo,
+      bool useNewLHCInfo);
+
+  void setFromLHCInfo(const LHCInfo& lhcInfo);
+  void setFromPerLS(const LHCInfoPerLS& infoPerLS);
+  void setFromPerFill(const LHCInfoPerFill& infoPerFill);
+
+  float crossingAngle();
+  static constexpr int crossingAngleInvalid = -1;
```
[the remark this answers is not in the document]
Comment (Glitchmin) [reply CondTools/RunInfo/interface/LHCInfoCombined.h:47]: It should be a float - my bad

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (Glitchmin) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:85]: Addressed in f5c24ce0440e40443b68d6b5158aa77dfc3229a2

File: CondTools/RunInfo/interface/LHCInfoCombined.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (Glitchmin) [reply CondTools/RunInfo/interface/LHCInfoCombined.h]: Addressed in f5c24ce0440e40443b68d6b5158aa77dfc3229a2

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc (hunk 13 above)
[the remark this answers is not in the document]
Comment (Glitchmin) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:85]: first part of the comment addressed in e7abd9ed1873a7bd85e5ec963e2ce9608e1f91be

**Comment by emanueleusai** [comment]
please test

**Comment by Glitchmin** [comment]
The tests that are crashing:

* `CalibTracker/SiStripCommon/test/test_all_cfg.py`
* `CalibTracker/SiStripChannelGain/test/testSSTGain_PCL_FromRECO_cfg.py`

are crashing also in my local enviroment even without any of my commits, both of them with a similar output:

```
testSSTGain_PCL_FromRECO_cfg.py:26: UserWarning: DAS query status: error
  warnings.warn("DAS query status: %s"%(status))
testSSTGain_PCL_FromRECO_cfg.py:93: UserWarning: ** WARNING: ** According to a DAS query no suitable data for test is available. Skipping test!
  warnings.warn("** WARNING: ** According to a DAS query no suitable data for test is available. Skipping test!")
  ```
  I'm not sure how to run `checkMultiRunHarvestingOutput` but it seems to be a part of `testSSTGainPCL_fromRECO`

**Comment by Glitchmin** [comment]
The tests that are crashing:

* `CalibTracker/SiStripCommon/test/test_all_cfg.py`
* `CalibTracker/SiStripChannelGain/test/testSSTGain_PCL_FromRECO_cfg.py`

are crashing also in my local enviroment even without any of my commits, both of them with a similar output:

```
testSSTGain_PCL_FromRECO_cfg.py:26: UserWarning: DAS query status: error
  warnings.warn("DAS query status: %s"%(status))
testSSTGain_PCL_FromRECO_cfg.py:93: UserWarning: ** WARNING: ** According to a DAS query no suitable data for test is available. Skipping test!
  warnings.warn("** WARNING: ** According to a DAS query no suitable data for test is available. Skipping test!")
  ```
  I'm not sure how to run `checkMultiRunHarvestingOutput` but it seems to be a part of `testSSTGainPCL_fromRECO`

**Comment by perrotta** [comment]
@Glitchmin the same unit tests are also crashing in the IB, see https://cmssdt.cern.ch/SDT/cgi-bin/showBuildLogs.py/el8_amd64_gcc11/www/thu/13.3-thu-23/CMSSW_13_3_X_2023-08-17-2300?utests

There is a fix ready to be merged for them, #42597

In the meanwhile, just ignore them.

Please finish applying the requested updates, namely applying the modifier only once into the module definition and not every time the module is colled. Then it would be great if you could squash all commits into one at the end, so that to get rid of the files of invalid states (see https://github.com/cms-sw/cmssw/pull/42515#issuecomment-1682558634) and clean up the history (which is helpful for possible future developers)

**Comment by Glitchmin** [comment]
I hope this commit addresses requested updates
I tested it using `runTheMatrix.py -l 1043,1044,1045 -j 8 --ibeos` and running `TestCTPPSDirectProtonSimulation`
Should I wait with squashing so those commits can be tested and commented?

**Comment by tvami** [comment]
hi @Glitchmin please squash and then we can trigger tests. Thanks!

**Comment by Glitchmin** [comment]
@tvami squashed and ready for testing

**Comment by tvami** [comment]
@cmsbuild , please test

**Comment by tvami** [comment]
@cmsbuild please test with https://github.com/cms-sw/cmssw/pull/42597

**Comment by vavati** [comment]
any further input/action required on our side?

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc [hunk 15]
```diff
@@ -61,26 +72,29 @@ std::shared_ptr<LHCInterpolatedOpticalFunctionsSetCollection> CTPPSInterpolatedO
...
```
Comment (perrotta) [inline CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:85]: If I understand it correctly, the crossing angle cannot be negative.
If so, to avoid the risk of requesting an exact comparisons between two floats (which can differ each other only because of some numerical fluctuation) you could just check instead:
```suggestion
  if (lhcInfoCombined.crossingAngle() < 0.f) {
```
or even
```suggestion
  if (lhcInfoCombined.crossingAngle() < -0.5f) {
```

File: Validation/CTPPS/plugins/CTPPSLHCInfoPlotter.cc [hunk 16]
```diff
@@ -83,14 +92,17 @@ void CTPPSLHCInfoPlotter::fillDescriptions(edm::ConfigurationDescriptions &descr
...
+  LHCInfoCombined lhcInfoCombined(iSetup, lhcInfoPerLSToken_, lhcInfoPerFillToken_, lhcInfoToken_, useNewLHCInfo_);
 
-  h_beamEnergy_->Fill(lhcInfo.energy());
-  h_xangle_->Fill(lhcInfo.crossingAngle());
-  h_betaStar_->Fill(lhcInfo.betaStar());
-  h2_betaStar_vs_xangle_->Fill(lhcInfo.crossingAngle(), lhcInfo.betaStar());
+  h_beamEnergy_->Fill(lhcInfoCombined.energy);
+  h_xangle_->Fill(lhcInfoCombined.crossingAngle());
+  h_betaStar_->Fill(lhcInfoCombined.betaStarX);  //adjust accordingly to run period
+  // h_betaStar_->Fill(lhcInfoCombined.betaStarY);
```
[the remark this answers is not in the document]
Comment (vavati) [reply Validation/CTPPS/plugins/CTPPSLHCInfoPlotter.cc:100]: Right. This was changed just to make the associated unit test working. There are several modules in Validation/ which require update of the LHCInfo and it will be done in another PR. Here in particular we should add a 2nd beta* definition: while the crossing angle is only in one direction (only X or Y), with new optics beta_x and beta_y are not anymore the same (this is new and never done till now)

File: RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc [hunk 17]
```diff
@@ -299,14 +318,18 @@ void CTPPSProtonProducer::produce(edm::Event &iEvent, const edm::EventSetup &iSe
 
               bool matching = true;
 
-              if (!ac.isSatisfied(ac.qX, tr_i.x(), tr_i.y(), hLHCInfo->crossingAngle(), tr_i.x() - tr_j.x()))
-                matching = false;
-              else if (!ac.isSatisfied(ac.qY, tr_i.x(), tr_i.y(), hLHCInfo->crossingAngle(), tr_i.y() - tr_j.y()))
+              if (!ac.isSatisfied(ac.qX, tr_i.x(), tr_i.y(), lhcInfoCombined.crossingAngleX, tr_i.x() - tr_j.x()))
```
[the remark this answers is not in the document]
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc:321]: You're right, I will push a commit addressing this soon

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc (hunk 15 above)
Comment (Glitchmin) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:85]: @vavati says that `crossingAngle` was never negative but from next year on it will be possible, so checking if xangle < 0.f is not a good idea.
 -1 as `crossingAngleInvalid` isn't bad since if `crossingAngle` is negative and valid it won't greater than around -140

File: RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc (hunk 17 above)
[the remark this answers is not in the document]
Comment (Glitchmin) [reply RecoPPS/ProtonReconstruction/plugins/CTPPSProtonProducer.cc]: addressed in 424c0b8c5d566c5cbc70a42cc0363eb9fbaf6cda

File: CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc (hunk 15 above)
Comment (Glitchmin) [reply CalibPPS/ESProducers/plugins/CTPPSInterpolatedOpticalFunctionsESSource.cc:85]: I do agree that comparing floats might be problematic - maybe I should introduce a static bool method to LHCInfoCombined named isCrossingAngleValid? It would simply look like this:
```
bool isCrossingAngleValid (float crossingAngle){
  return !(crossingAngle >= crossingAngleInvalid-1e-6 && crossingAngle <= crossingAngleInvalid+1e-6)
}
```

**Comment by perrotta** [comment]
crossingAngleX and `crossingAngleY` (and probably also something else) should become private members of the class, to be accessed only through the accessors.
There is no need to implement it now: also in a follow up PR, unless you want to modify this one further anyhow, e.g. to implement https://github.com/cms-sw/cmssw/pull/42515#discussion_r1301607066

**Comment by Glitchmin** [comment]
> 

For now we're not sure if we won't need to use `crossingAngleX` and `crossingAngleY` seperately like here: https://github.com/cms-sw/cmssw/blob/d62c8f53a57ff413760a3df1225dfb760efb24a2/CalibPPS/ESProducers/plugins/CTPPSBeamParametersFromLHCInfoESSource.cc#L81 so it's best to leave them public

**Comment by Glitchmin** [comment]
I pushed a new commit being a squash of a previous commit and 2 new:
*one replacing usages of `crossingAngleX` with `crossingAngle()`
*second introducing `isCrossingAngleInvalid()` method that avoids direct float comparison

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by Glitchmin** [comment]
> @cmsbuild please test with #42597

@perrotta can you please add PR fixing `testCalibTrackerSiStripCommon` tests?

**Comment by perrotta** [comment]
> @perrotta can you please add PR fixing `testCalibTrackerSiStripCommon` tests?

It is merged in master since one week already

**Comment by Glitchmin** [comment]
> > @perrotta can you please add PR fixing `testCalibTrackerSiStripCommon` tests?
> 
> It is merged in master since one week already

My branch isn't up to date with master: https://github.com/CTPPS/cmssw/tree/ak_new_lhcInfo_reco_squashed2
Will the tests still be ok? I'm not sure if my PR is merged into master for tests or is it simply my branch that is tested

**Comment by perrotta** [comment]
@Glitchmin this PR has to be merged in the master, and therefore it is tested with the latest produced IB. Since the 1100 is not yet available, it will be CMSSW_13_3_X_2023-08-24-2300 I guess.

**Comment by emanueleusai** [comment]
+1

**Comment by mandrenguyen** [comment]
+1

**Comment by Glitchmin** [comment]
From my side I'm not planning any more updates, I'm looking forward to more comments or approval

**Comment by perrotta** [comment]
+alca

**Comment by perrotta** [comment]
+db

**Comment by perrotta** [comment]
+1
